sprinkle/low-level/context.py

```diff
@@ -0,0 +1,127 @@
+
+ - The same context object must be usable in both local training and cluster
+   training environments.  Obviously if it exposes cluster-training-specific
+   extensions during cluster training, those extensions do not need to work in
+   local training environments, but any user code which does not depend on
+   such environment-specific extensions should run equally well in either
+   environment.
+
+ - The same context object must be usable in both training and non-training
+   workloads.  The sprinkle API for training should feel analagous to the
+   sprnkle API for distributed batch inference and to distributed test set
+   metrics, even though the calls the user makes will differ.
+
+ - Backwards compatibility: the existing Trial APIs need to still work, so
+   those TrialContext classes should remain unchanged.
+"""
+
+# GENERIC CONTEXT
+# Useful for custom workloads and is the basis for the framework contexts.
+context = det.GenericContext()
+
+# distributed data: applies to any workload; training or inference, local or cluster.
+# (same as the DistributedContext right now)
+context.distributed.get_rank()
+context.distributed.get_size()
+context.distributed.get_local_rank()
+context.distributed.get_num_agents()
+
+# rendezvous data: info about machines and IPs that are participating in distributed jobs.
+# This is configured automatically by the rendezvous info layer in Determined, but can also be
+# configured automatically through an environment variable if you want to use Determined's dtrain
+# features outside of Determined.
+context.rendezvous.get_num_nodes()
+context.rendezvous.get_node_rank()
+context.rendezvous.get_slots_per_node()
+
+# master data: applies to any job running on-cluster, of any job that you have
+# chosen to connect to the master with context.connect_to_master()
+context.connect_to_master(...)  # args similar to Determined() object
+context.master.get_addr()
+context.master.get_port()
+context.master.get_tls()
+
+# cluster data: applies to any job running on-cluster
+context.cluster.get_container_id()
+context.cluster.get_container_gpus()
+context.cluster.get_slot_ids()
+
+# training-related data; only applies to `det e create` workloads
+context.training.get_experiment_config()
+context.training.get_data_config()
+context.training.get_experiment_id()
+context.training.get_trial_id()
+context.training.get_global_batch_size()
+context.training.get_per_slot_batch_size()
+context.training.get_hparams()
+context.training.get_trial_seed()
+context.training.get_latest_checkpoint()
+
+"""
+PUSH-ARCHITECTURE RELATED APIS:
+
+The core features of our platform can be delivered if we only have metrics and checkpoints for
+training.  Advanced features like preemption and hpsearch should also be possible, but MUST NOT
+be necessary to deliver basic features.
+"""
+
+# Training metrics
+context.training.begin_training()  # optional call, improves webui experience
```

Comment:
What does "create an active training metric" mean? You would already know how many batches you completed from before I think.

> And it would be weird to expose this function to users since it would add more mind effort to them.

No, it's fine.  It's optional.  If users want to tell the webui what they are doing, they can call it.  Or they can safely ignore it.